src/mito2/src/region/opener.rs

```diff
@@ -124,20 +133,72 @@ impl RegionOpener {
             }
         );
 
+        let region_id = metadata.region_id;
         let mutable = self.memtable_builder.build(&metadata);
         let version = VersionBuilder::new(metadata, mutable).build();
-        let version_control = Arc::new(VersionControl::new(version));
+        let flushed_sequence = version.flushed_sequence;
 
-        // TODO(yingwen): Replay.
+        let version_control = Arc::new(VersionControl::new(version));
+        replay_memtable(
+            &self.wal,
+            region_id,
+            flushed_sequence,
+            version_control.clone(),
+        )
+        .await?;
 
-        Ok(MitoRegion {
+        let region = MitoRegion {
             region_id: self.region_id,
             version_control,
             manifest_manager,
-        })
+        };
+        Ok(region)
     }
 }
 
+/// Replays the mutations from WAL and inserts mutations to memtable of given region.
+async fn replay_memtable<S: LogStore>(
+    wal: &Wal<S>,
+    region_id: RegionId,
+    flushed_sequence: SequenceNumber,
+    version_control: VersionControlRef,
+) -> Result<()> {
+    let mut notify_waiter = vec![];
+    {
+        let mut region_writer_ctx = RegionWriteCtx::new(region_id, version_control);
+        let mut wal_stream = wal.scan(region_id, flushed_sequence)?;
+        while let Some(res) = wal_stream.next().await {
+            let (_, entry) = res?;
+            for mutation in entry.mutations.into_iter() {
+                let (tx, rx) = tokio::sync::oneshot::channel();
+                region_writer_ctx.push_mutation(mutation.op_type, mutation.rows, Some(tx));
+                notify_waiter.push(rx);
+            }
+        }
+        region_writer_ctx.write_memtable();
+    }
+
+    let rows_replayed: usize = futures::future::try_join_all(notify_waiter)
+        .await
+        .context(RecvSnafu)?
+        .into_iter()
+        .collect::<Result<Vec<Output>>>()?
+        .into_iter()
+        .map(|output| {
+            let Output::AffectedRows(rows_inserted) = output else {
+                unreachable!("Only expect affected rows during WAL replay.")
+            };
+            rows_inserted
+        })
+        .sum();
```

Comment:
We can sum the inserted rows in the replay loop. Then receiver of the write notify is used to notify callers outside.